lib/tar.ml

```diff
@@ -62,45 +81,162 @@ module Header = struct
+      (Cstruct.to_string ~off:hdr_file_name_off ~len:sizeof_hdr_file_name buf)
+  let set_hdr_file_name buf v =
+    let v = marshal_string v sizeof_hdr_file_name in
+    Cstruct.blit_from_string v 0 buf hdr_file_name_off sizeof_hdr_file_name
+
+  let get_hdr_file_mode buf =
+    unmarshal_int
+      (Cstruct.to_string ~off:hdr_file_mode_off ~len:sizeof_hdr_file_mode buf)
+  let set_hdr_file_mode buf v =
+    let v = marshal_int v sizeof_hdr_file_mode in
+    Cstruct.blit_from_string v 0 buf hdr_file_mode_off sizeof_hdr_file_mode
+
+  let get_hdr_user_id buf =
+    unmarshal_int
+      (Cstruct.to_string ~off:hdr_user_id_off ~len:sizeof_hdr_user_id buf)
+  let set_hdr_user_id buf v =
+    let v = marshal_int v sizeof_hdr_user_id in
+    Cstruct.blit_from_string v 0 buf hdr_user_id_off sizeof_hdr_user_id
+
+  let get_hdr_group_id buf =
+    unmarshal_int
+      (Cstruct.to_string ~off:hdr_group_id_off ~len:sizeof_hdr_group_id buf)
+  let set_hdr_group_id buf v =
+    let v = marshal_int v sizeof_hdr_group_id in
+    Cstruct.blit_from_string v 0 buf hdr_group_id_off sizeof_hdr_group_id
+
+  let get_hdr_file_size buf =
+    unmarshal_int64
+      (Cstruct.to_string ~off:hdr_file_size_off ~len:sizeof_hdr_file_size buf)
+  let set_hdr_file_size buf v =
+    let v = marshal_int64 v sizeof_hdr_file_size in
+    Cstruct.blit_from_string v 0 buf hdr_file_size_off sizeof_hdr_file_size
+
+  let get_hdr_mod_time buf =
+    unmarshal_int64
+      (Cstruct.to_string ~off:hdr_mod_time_off ~len:sizeof_hdr_mod_time buf)
+  let set_hdr_mod_time buf v =
+    let v = marshal_int64 v sizeof_hdr_mod_time in
+    Cstruct.blit_from_string v 0 buf hdr_mod_time_off sizeof_hdr_mod_time
+
+  let get_hdr_chksum buf =
+    unmarshal_int64
+      (Cstruct.to_string ~off:hdr_chksum_off ~len:sizeof_hdr_chksum buf)
+  let set_hdr_chksum buf v =
+    let v = marshal_int64 v sizeof_hdr_chksum in
+    Cstruct.blit_from_string v 0 buf hdr_chksum_off sizeof_hdr_chksum
+
+  let get_hdr_link_indicator buf = Cstruct.get_char buf hdr_link_indicator_off
+  let set_hdr_link_indicator buf v =
+    Cstruct.set_char buf hdr_link_indicator_off v
+
+  let get_hdr_link_name buf =
+    unmarshal_string
+      (Cstruct.to_string ~off:hdr_link_name_off ~len:sizeof_hdr_link_name buf)
+  let set_hdr_link_name buf v =
+    let v = marshal_string v sizeof_hdr_link_name in
+    Cstruct.blit_from_string v 0 buf hdr_link_name_off sizeof_hdr_link_name
+
+  let get_hdr_magic buf =
+    unmarshal_string
+      (Cstruct.to_string ~off:hdr_magic_off ~len:sizeof_hdr_magic buf)
+  let set_hdr_magic buf v =
+    let v = marshal_string v sizeof_hdr_magic in
+    Cstruct.blit_from_string v 0 buf hdr_magic_off sizeof_hdr_magic
+
+  let _get_hdr_version buf =
+    unmarshal_int
+      (Cstruct.to_string ~off:hdr_version_off ~len:sizeof_hdr_version buf)
+  let set_hdr_version buf v =
+    let v = marshal_int v sizeof_hdr_version in
+    Cstruct.blit_from_string v 0 buf hdr_version_off sizeof_hdr_version
```

Comment:
The version field is a bit special. In case of GNU we marshal the string `" "` and write that. In case of ustar we write `"00"` (though for a long time we incorrectly wrote `"0\000"` without complaints).
```suggestion
  let _get_hdr_version buf =
    Cstruct.to_string ~off:hdr_version_off ~len:sizeof_hdr_version buf
  let set_hdr_version buf v =
    Cstruct.blit_from_string v 0 buf hdr_version_off sizeof_hdr_version
```